socorro/lib/javautil.py

```diff
@@ -0,0 +1,113 @@
+        self.stack = stack
+        self.additional = additional
+
+    def to_public_string(self):
+        """Builds stack trace with parts of the exception that don't contain PII
+
+        Currently, this skips the exception message which can contain data
+        related to the exception and might have PII in it.
+
+        It also skips the "Suppressed" and "Caused by" sections because those
+        can have exception messages, too.
+
+        """
+        text = '%s\n%s' % (
+            self.exception_class,
+            '\n'.join([
+                '\t%s' % line for line in self.stack
+            ])
+        )
+        return text
+
+
+class MalformedJavaStackTrace(Exception):
+    pass
+
+
+CLASS_MESSAGE, STACK, ADDITIONAL = range(3)
+
+
+def parse_java_stack_trace(text):
+    """Parses a JavaStackTrace text blob into a JavaException instance
+
+    :arg str text: the JavaStackTrace blob
+
+    :returns: a JavaException instance
+
+    :raises MalformedJavaStackTrace: if there's a parsing error
+
+    """
+    if not text:
+        raise MalformedJavaStackTrace('no text blob')
+
+    stage = CLASS_MESSAGE
+    lines = peekable(text.splitlines())
+
+    new_exc = JavaException(
+        exception_class='',
+        exception_message='',
+        stack=[],
+        additional=[]
+    )
+
+    for line in lines:
+        if not line.strip():
+            continue
+
+        if stage is CLASS_MESSAGE:
+            try:
+                cls, msg = line.split(':', 1)
+            except ValueError as ve:
+                # If there's no : in the line, then it raises a ValueError.
+                # That means this is malformed.
+                raise MalformedJavaStackTrace(repr(ve))
+
+            # Append lines to the message until one of them starts with
+            # a tab.
+            next_line = lines.peek(None)
+            while next_line and not next_line.startswith('\tat '):
+                msg = msg + '\n' + lines.next()
+                next_line = lines.peek(None)
+
+            new_exc.exception_class = cls.strip()
+            new_exc.exception_message = msg.strip()
+
+            stage = STACK
+
+        elif stage is STACK:
+            # Verify this has a tab at the beginning of the line
+            if line[0] != '\t':
+                raise MalformedJavaStackTrace('stack line missing tab')
+
+            # Drop first tab from the line
+            new_exc.stack.append(line[1:])
+
+            next_line = lines.peek(None)
+            if next_line and next_line.strip().startswith(('Suppressed:', 'Caused by:')):
+                stage = ADDITIONAL
+
+        elif stage is ADDITIONAL:
+            # Verify this has a tab at the beginning of the line
+            if line[0] != '\t':
+                raise MalformedJavaStackTrace('additional line missing tab')
+
+            # Drop first tab from the line
+            new_exc.additional.append(line[1:])
+
+    return new_exc
```

Comment:
Not the best DFA I've ever coded, but I was in a rush so I coded for readability and reasonability figuring those are the most important properties to have at the moment.